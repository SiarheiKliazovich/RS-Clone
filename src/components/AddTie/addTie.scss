@import "../../sass/mixins";
@import "../../sass/variables";

.tieaddform__wrapper {
  background: #f6f6f6;
}

.form-addtie__title {
  text-align: center;
  font-size: $font-size-xlg;
  font-family: $font-family-base;
  padding-top: 15px;
}

.tieaddform__banner-wrapper {
  background: #f6f6f6;
  padding-top: 40px;
  padding-bottom: 40px;
}

.tieaddform-block {
  height: calc(100% - 30px);
  border: none;
  background: #ffffff;
}

.form-addtie__wrapper {
  display: flex;
  justify-content: center;
  align-items: center;
  margin-left: 20px;
}

.form-addtie__content {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  width: 60%;
}

.form-img-blok {
  background-image: url(../../assets/images/png/loade.png);
  background-repeat: no-repeat;
  background-size: 100%;
  width: 300px;
  height: 205px;

    @include media-small-mobile {
      display: none;
    }
}

.tieadd-form {
  display: flex;
  flex-direction: column;
}

.tieadd-form__label {
  padding-top: 15px;
  font-size: $font-size-llg;
  font-family: $font-family-base;
}

.field-file {
  margin-top: 15px;
  font-size: 140%;
}

.success-tooltip {
  font-family: $font-family-base;
  padding-bottom: 15px;
}

.add-tie-btn {
  border: none;
  border-radius: 3px;
  color: $color-white1;
  transition: 0.4s ease-in-out;
  padding: 10px;
  cursor: pointer;
  max-width: 200px;
  text-align: left;
  margin: 20px 0;

  .fa {
    margin-right: 10px;
    color: $color-white1;
  }
}
