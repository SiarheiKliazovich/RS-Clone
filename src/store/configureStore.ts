/* eslint-disable @typescript-eslint/ban-ts-comment */
import { applyMiddleware, createStore } from "redux";
import thunkMiddleware from "redux-thunk";

import loggerMiddleware from "./logger";
import rootReducer from "../reducers";

const configureStore = () => {
  const middlewares = [loggerMiddleware, thunkMiddleware];
  const middlewareEnhancer = applyMiddleware(...middlewares);

  const store = createStore(rootReducer, middlewareEnhancer);

  // @ts-ignore:next-line
  if (process.env.NODE_ENV !== "production" && module.hot) {
    // @ts-ignore:next-line
    module.hot.accept("./../reducers", () => store.replaceReducer(rootReducer));
  }

  return store;
};

export default configureStore;
