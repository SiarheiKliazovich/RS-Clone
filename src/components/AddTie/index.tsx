import { FunctionComponent, useState, ChangeEvent } from "react";
import { useSelector, useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import * as Yup from "yup";
import { Formik, Field, Form, ErrorMessage } from "formik";

import "./addTie.scss";
import Hover from "../Hover";
import { IStore } from "../../interfaces/store";
import { nightTheme } from "../../data/constants";
import addTie from "../../data/addTie";
import { createTie } from "../../services/apiTies";
import { ITie } from "../../interfaces/tie";
import { ILangReducer } from "../../interfaces/langReducer";
import {
  addTieFetching,
  addTieFetched,
  addTieFetchingError,
} from "../../actions/addTie";
import { IAddTieReducer } from "../../interfaces/addTie";
import { convertBase64 } from "../../helpers/convertBase64";
import Spinner from "../Spinner";

const AddTie: FunctionComponent = () => {
  const interfaceSettings = useSelector((state: IStore) => state.appInterface);
  const { accentColor, isNavbarNightMode } = interfaceSettings;
  const backgroundColor = nightTheme.background.element;
  const { lang } = useSelector((state: ILangReducer) => state.langReducer);
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const list = addTie.find((c) => c.lang === lang)!;
  const [base64Image, setBase64Image] = useState<string>("");
  const [imageError, setImageError] = useState<boolean>(false);
  const dispatch = useDispatch();
  const addTiesStore = useSelector(
    (state: IAddTieReducer) => state.addTieReducer
  );
  const { addTieLoadingStatus } = addTiesStore;

  const uploadImage = async (event: ChangeEvent<HTMLInputElement>) => {
    setImageError(false);
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const file = event!.target!.files![0]!;
    const base64 = (await convertBase64(file)) as string;
    setBase64Image(base64);
  };

  const createTieForm = async (formData: Pick<ITie, "name" | "price">) => {
    setImageError(!base64Image);
    if (!imageError) {
      try {
        dispatch(addTieFetching());
        const login = localStorage.getItem("login");
        const user = login ? JSON.parse(login) : [];
        const body = {
          ...formData,
          userId: user.user._id,
          image: base64Image,
        };

        const resp = await createTie(body);
        if (resp.userId) {
          dispatch(addTieFetched(resp));
        } else {
          dispatch(addTieFetchingError());
        }
      } catch (e) {
        dispatch(addTieFetchingError());
      }
    }
  };
  const isLoading = addTieLoadingStatus === "loading";

  return (
    <main className="main-market">
      <div className="tieaddform__wrapper">
        <div className="container">
          <div className="tieaddform__banner-wrapper">
            <div className="tieaddform-block">
              <div className="form-addtie__title">{list.data.titleblock}</div>
              <div className="form-addtie__wrapper">
                <div className="form-addtie__content">
                  {isLoading && <Spinner />}
                  {!isLoading && (
                    <Formik
                      initialValues={{
                        name: "",
                        price: 0,
                      }}
                      validationSchema={Yup.object({
                        name: Yup.string().required(`${list.data.titleerror}`),
                        price: Yup.number()
                          .moreThan(0)
                          .required(`${list.data.priceerror}`),
                      })}
                      onSubmit={(values) => {
                        createTieForm(values);
                      }}
                    >
                      <Form className="tieadd-form">
                        <label htmlFor="name" className="tieadd-form__label">
                          {list.data.title}
                        </label>
                        <Field
                          className="field"
                          id="name"
                          name="name"
                          type="name"
                          placeholder="Title"
                        />
                        <ErrorMessage
                          className="error"
                          name="name"
                          component="div"
                        />
                        <label htmlFor="price" className="tieadd-form__label">
                          {list.data.pricetitle}
                        </label>
                        <Field
                          className="field"
                          id="price"
                          name="price"
                          type="number"
                          placeholder="Price"
                        />
                        <ErrorMessage
                          className="error"
                          name="price"
                          component="div"
                        />
                        <label htmlFor="image" className="tieadd-form__label">
                          {list.data.picturetitle}
                        </label>
                        <input
                          className="field-file"
                          id="image"
                          name="image"
                          type="file"
                          placeholder="image"
                          onChange={uploadImage}
                        />
                        {imageError && (
                          <p className="error">{list.data.pictureerror}</p>
                        )}
                        <Hover>
                          <button
                            style={{
                              backgroundColor: isNavbarNightMode
                                ? backgroundColor
                                : accentColor.static,
                            }}
                            className="add-tie-btn"
                            type="submit"
                          >
                            <i className="fa fa-align-justify" />
                            {list.data.btn}
                          </button>
                        </Hover>
                      </Form>
                    </Formik>
                  )}
                  {addTieLoadingStatus === "error" && (
                    <div className="error-tooltip">
                      <i className="fa fa-warning" />
                      {list.data.errortooltip}
                    </div>
                  )}
                  {addTieLoadingStatus === "loaded" && (
                    <div className="success-tooltip">
                      <i className="fa fa-info" />
                      {list.data.tooltip}{" "}
                      <Link to="/tie-market">{list.data.tiemarket}</Link>
                      {list.data.or}{" "}
                      <Link to="/my-ties">{list.data.myties}</Link>
                    </div>
                  )}
                </div>
                <div className="form-img-blok"></div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </main>
  );
};
export default AddTie;
