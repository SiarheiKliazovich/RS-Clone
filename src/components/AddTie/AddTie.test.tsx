import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import configureStore from "redux-mock-store";
import AddTie from ".";

const mockStore = configureStore([]);

describe("AddTie ", () => {
  // eslint-disable-next-line
  let store: any;

  beforeEach(() => {
    store = mockStore({
      appInterface: {
        accentColor: {
          static: "#000",
          hover: "#fff",
        },
      },
      langReducer: {
        lang: "en",
      },
      addTieReducer: {
        addTieLoadingStatus: "load",
      },
    });
  });

  test("renders correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <AddTie />
        </Router>
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
