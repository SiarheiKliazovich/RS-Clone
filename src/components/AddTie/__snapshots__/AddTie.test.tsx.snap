// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTie  renders correctly 1`] = `
<main
  class="main-market"
>
  <div
    class="tieaddform__wrapper"
    style="background-color: rgb(246, 246, 246);"
  >
    <div
      class="container"
    >
      <div
        class="tieaddform__banner-wrapper"
        style="background-color: rgb(246, 246, 246);"
      >
        <div
          class="tieaddform-block"
          style="background-color: rgb(255, 255, 255);"
        >
          <div
            class="form-addtie__title"
          >
            Create a tie
          </div>
          <div
            class="form-addtie__wrapper"
          >
            <div
              class="form-addtie__content"
            >
              <form
                action="#"
                class="tieadd-form"
              >
                <label
                  class="tieadd-form__label"
                  for="name"
                >
                  Add tie title
                </label>
                <input
                  class="field"
                  id="name"
                  name="name"
                  placeholder="Title"
                  type="name"
                  value=""
                />
                <label
                  class="tieadd-form__label"
                  for="price"
                >
                  Add tie price:
                </label>
                <input
                  class="field"
                  id="price"
                  name="price"
                  placeholder="Price"
                  type="number"
                  value="0"
                />
                <label
                  class="tieadd-form__label"
                  for="image"
                >
                  Upload tie picture
                </label>
                <input
                  class="field-file"
                  id="image"
                  name="image"
                  placeholder="image"
                  type="file"
                />
                <button
                  class="add-tie-btn"
                  style="background-color: rgb(0, 0, 0);"
                  type="submit"
                >
                  <i
                    class="fa fa-align-justify"
                  />
                  Create tie
                </button>
              </form>
            </div>
            <div
              class="form-img-blok"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</main>
`;
